Python/_warnings.c

```diff
@@ -443,6 +491,174 @@ warn_explicit(PyObject *category, PyObject *message,
+{
+    PyObject *key = NULL, *fix_txt = NULL, *text = NULL, *result = NULL, *lineno_obj = NULL;
+    PyObject *item = Py_None;
+    const char *action;
+    int rc, rc_msg, rc_fix;
+
+    if (registry && !PyDict_Check(registry) && (registry != Py_None)) {
+        PyErr_SetString(PyExc_TypeError, "'registry' must be a dict or None");
+        return NULL;
+    }
+
+    /* Normalize module. */
+    if (module == NULL) {
+        module = normalize_module(filename);
+        if (module == NULL)
+            return NULL;
+    }
+    else
+        Py_INCREF(module);
+
+    /* Normalize message and fix. */
+    Py_INCREF(message);  /* DECREF'ed in cleanup. */
+    Py_INCREF(fix);  /* DECREF'ed in cleanup. */
+    rc_msg = PyObject_IsInstance(message, PyExc_Warning);
+    rc_fix = PyObject_IsInstance(fix, PyExc_Warning);
+    if (rc_msg == -1 || rc_fix == -1) {
+        goto cleanup;
+    }
+    if (rc_msg == 1) {
+        text = PyObject_Str(message);
+        if (text == NULL)
+            goto cleanup;
+        category = (PyObject*)message->ob_type;
+    }
+    else {
+        text = message;
+        message = PyObject_CallFunction(category, "O", message);
+        if (message == NULL)
+            goto cleanup;
+    }
+
+    if (rc_fix == 1) {
+        fix_txt = PyObject_Str(fix);
+        if (fix_txt == NULL)
+            goto cleanup;
+        category = (PyObject*)fix->ob_type;
+    }
+    else {
+        fix_txt = fix;
+        fix = PyObject_CallFunction(category, "O", fix);
+        if (fix == NULL)
+            goto cleanup;
+    }
+
+    lineno_obj = PyInt_FromLong(lineno);
+    if (lineno_obj == NULL)
+        goto cleanup;
+
+    /* Create key. */
+    key = PyTuple_Pack(4, text, fix_txt, category, lineno_obj);
+    if (key == NULL)
+        goto cleanup;
+
+    if ((registry != NULL) && (registry != Py_None)) {
+        rc = already_warned(registry, key, 0);
+        if (rc == -1)
+            goto cleanup;
+        else if (rc == 1)
+            goto return_none;
+        /* Else this warning hasn't been generated before. */
+    }
+
+    action = get_filter(category, text, lineno, module, &item);
+    if (action == NULL)
+        goto cleanup;
+
+    if (strcmp(action, "error") == 0) {
+        PyErr_SetObject(category, message);
+        goto cleanup;
+    }
+
+    /* Store in the registry that we've been here, *except* when the action
+       is "always". */
+    rc = 0;
+    if (strcmp(action, "always") != 0) {
+        if (registry != NULL && registry != Py_None &&
+                PyDict_SetItem(registry, key, Py_True) < 0)
+            goto cleanup;
+        else if (strcmp(action, "ignore") == 0)
+            goto return_none;
+        else if (strcmp(action, "once") == 0) {
+            if (registry == NULL || registry == Py_None) {
+                registry = get_once_registry();
+                if (registry == NULL)
+                    goto cleanup;
+            }
+            /* _once_registry[(text, category)] = 1 */
```

Comment:
In general, we try not to commit commented code. I *think* this line can be safely deleted (as can line 598).